### PR TITLE
Issue/2277 typedmethod return optional

### LINKED ISSUE #2277: @typedmethod does not allow Optional
Results in `TypeError: issubclass() arg 1 must be a class` on `File "src/inmanta/protocol/common.py", line 504, in _validate_return_type`.

### LINKED ISSUE #2279: make get_compile_data a @typedmethod

### PR DESCRIPTION
**Pull request by sanderr**
# Description

* Short description here *

closes #2277 
closes #2279 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] ~~End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/protocol/common.py [hunk 1]
```diff
@@ -505,14 +505,14 @@ def _validate_return_type(self, arg_type: Type) -> None:
...
-        elif not typing_inspect.is_generic_type(arg_type) and issubclass(arg_type, ReturnValue):
-            raise InvalidMethodDefinition("ReturnValue should have a type specified.")
-
-        elif not typing_inspect.is_generic_type(arg_type) and issubclass(arg_type, type(None)):
-            pass
+        elif not typing_inspect.is_generic_type(arg_type) and isinstance(arg_type, type):
+            if issubclass(arg_type, ReturnValue):
+                raise InvalidMethodDefinition("ReturnValue should have a type specified.")
 
+            elif issubclass(arg_type, ReturnValue):
```
Comment (bartv) [inline src/inmanta/protocol/common.py:512]: Why is this here? This is exactly the same as the if condition

Comment (sanderr) [reply src/inmanta/protocol/common.py:512]: Good catch, this should be issubclass(arg_type, type(None)). I accidentally changed it.

Comment (sanderr) [reply src/inmanta/protocol/common.py:512]: Though actually this is redundant with `allow_none_type=True`, I'm gonna leave it out.

File: tests/test_server.py [hunk 2]
```diff
@@ -962,3 +962,13 @@ async def test_resource_action_pagination(postgresql_client, client, clienthelpe
     response = json.loads(response.body.decode("utf-8"))
     action_ids = [uuid.UUID(resource_action["action_id"]) for resource_action in response["data"]]
     assert action_ids == third_page_action_ids
+
+
+def test_2277_typedmethod_return_optional() -> None:
```
Comment (wouterdb) [inline tests/test_server.py:967]: best to also test that the None is correctly encoded/decoded

Comment (sanderr) [reply tests/test_server.py:967]: In the current implementation `{}` is returned instead of `{"data": null}`. I could easily change that but then `tests/test_protocol::test_return_non_warnings` fails because it asserts `"data not in response.result`. Is this assertion now obsolete, or do we observe two distinct cases, one where "data" should be absent and one where it should contain `null`?

File: tests/test_protocol.py [hunk 3]
```diff
@@ -1590,3 +1592,88 @@ async def test_methodY(self, id: str, name: str, age: int) -> str:  # NOQA
 
     request = MethodProperties.methods["test_method"][0].build_call(args=[], kwargs={"id": "1", "name": "monty", "age": 42})
     assert request.url == "/api/v1/test/1/monty?age=42"
+
+
+@pytest.mark.asyncio(timeout=5)
+async def test_2151_method_header_parameter_in_body(async_finalizer) -> None:
```
Comment (sanderr) [inline tests/test_protocol.py:1598]: I've moved this to `tests/test_protocol.py` because that seems to be the correct place. I had added this to `test_server.py` a couple of weeks ago.

File: tests/test_server.py (hunk 2 above)
Comment (sanderr) [reply tests/test_server.py]: @wouterdb What is your opinion on this? Should I update this assertion in the failing test?

Comment (wouterdb) [reply tests/test_server.py]: https://jsonapi.org/format/#auto-id--top-level 

it should be null to be conform the spec, not absent. 

I think that testcase can be updated. 

(if meta is present, it data could be absent)
